### DISCUSSION TITLE
How to use Dagster Pipes along with AWS StepFunctions?

### DISCUSSION THREAD

**Discussion by mraronsimon** [dagster-io/dagster#19558] [Q&A]
Hi Dagster Team!

We want to integrate Dagster into our existing AWS environment. Hosting Dagster on AWS sounds easy (thanks for the great documentation), but **we already have some predefined jobs (mostly in AWS StepFunction) that we wouldn't change** (or just a bit). 
_Additionally, **StepFunction integration** could make a **great opportunity to integrate Dagster with most of the AWS services** (e.g. Athena) and gives the **ability to define complex jobs (for a single Asset creation) involving multiple AWS services** in a strict flow._

I started to discover how to integrate our StepFunction jobs by Dagster's Pipes, however, I got stuck.
**I have a TODO list:**

1. I should implement a new pipe client that could start StepFunction through AWS SDK _(sounds easy, skip for now)_. 
2. I should use dagster_pipes Python package in StepFunction's steps (Lambda/Glue) _(that could be a problem, **focus on that!**)_

I imagine, my StepFunction contains **multiple "steps"** which will use `open_dagster_pipes`. Most of the steps (internal steps) will not report asset materialization (just the termination steps). Each step would use the same pipe context and send messages to the same pipes' session. 

The problem is that the "with" statement of `open_dagster_pipes` will **close the session after we leave it** _(but we would like to reopen it in the next step of our StepFunction job/pipeline)_.

 Instead of using AWS StepFunction, I tried to **reproduce that behavior by a simple** `pipes_subprocess_client`.

```python
# orchestration
@asset()
def subprocess_asset(
    context: AssetExecutionContext, pipes_subprocess_client: PipesSubprocessClient
):
    cmd = [
        shutil.which("python"),
        file_relative_path(__file__, "external_code.py"),
    ]
    return pipes_subprocess_client.run(
        command=cmd, context=context
    ).get_materialize_result()

defs = Definitions(
    assets=[subprocess_asset],
    resources={"pipes_subprocess_client": PipesSubprocessClient()},
)
```

```python
# external job

def first_step():
    # AWS Lambda/Glue body start
    with open_dagster_pipes():
        context = PipesContext.get()
        context.log.debug("This is my first step.")
        context.log.debug("Sorry I can't report the asset, I responsible of a partial problem.")
    # AWS Lambda/Glue body end


def next_step():
    with open_dagster_pipes():
        context = PipesContext.get()
        context.log.debug("Here we are! I'm the next step. :)")
        context.log.debug("Sorry I can't report the asset, I responsible of a partial problem.")


def final_step():
    with open_dagster_pipes():
        context = PipesContext.get()
        context.log.debug("Here we are! I'm the final step. :)")
        context.log.debug("The previous 2 AWS Lambda done with some preparation, I will report the asset.")
        # send structured metadata back to Dagster
        context.report_asset_materialization(metadata={"total_orders": [1,2,3]})

def main():
    # Insted of AWS StepFunction It will emulate multiple AWS Lambda function invocation
    first_step() # AWS Lambda or Glue
    time.sleep(3) # Step function start the next step
    next_step() # AWS Lambda or Glue
    time.sleep(3) # Step function start the final step
    final_step() # AWS Lambda or Glue

if __name__ == "__main__":
    main()
```

**Could you help a bit?**
Perhaps you have an existing article/discussion about that problem, I just didn't find it.
Thanks for your answers!

Best regards,
Aron Simon

(FYI @l1x)

**Comment by alangenfeld** [comment]
The initial design for pipes expects that there is a 1:1 relationship between the orchestration context and a single external process which will use that one pipes session for its full life. Supporting repeated open/close in the message handler is likely a small tweak, but updating the message reader/writers to work correctly in this setup will be a bigger lift (especially if we want to handle concurrency).

**Reply by mraronsimon** [reply]
**Thank you for your quick answer!**

We will examine within the team how committed we are to the AWS StepFunctions. However, we are also able to operate with ExternalAssets if I'm correct (but triggering the StepFunctions will be less elegant).

**Reply by mraronsimon** [reply]
**Another question** _(related to the previous topic)_:

What if I initiate an **internal Dagster Python job/asset** that **controls StepFunctions** _(launches it and polls its state in a custom way // without using Pipes)_ and, if necessary, **collects logs** related to the execution **from CloudWatch**? 

Is this a feasible approach? What is your opinion?

**Reply by alangenfeld** [reply]
> able to operate with ExternalAssets if I'm correct 

Yep, you could declare some external assets and then use the "report" APIs from StepFunctions to inform Dagster of relevant materializations. 

> Is this a feasible approach? What is your opinion?

I see no reason why it wouldn't be feasible. CloudWatch can be a bit annoying, but I expect the core kick-off and poll for StepFunctions part to be pretty straight forward.

**Comment by tomsej** [comment]
Hi @mraronsimon, what an interesting discussion! Have you come across any solutions yet? We're also heavily reliant on StepFunctions (I find it particularly appealing as it acts as an interface for execution types. For instance, you can switch between AWS Lambda and AWS Batch, yet the SFN arguments remain consistent). Currently we are using ECS Run Launcher and Multiprocess Step Executor. While we've created a simple resource for executing SFN, we're presently not retrieving any logs from CloudWatch (it's something on our backlog). It works, but we need to use a beefy ECS instance to run multiple steps in parallel, as each step operates as a separate process, solely for executing SFN and polling for state (what is a huge waste of resources). This is an issue that could certainly be resolved by an async/threaded #4041 step executor. I'm unsure if Pipes would offer a solution. Hence, I'm contemplating whether we should experiment with pipes or develop our custom step executor...

**Reply by mraronsimon** [reply]
Hi @tomsej! Interesting tools, interesting questions! Unfortunately for me, this was just a brief task, as I have a different role in the project, so I won't be able to provide answers due to I don't have much experience.

I integrated the pipes functionality for AWS Glue as well (multiple executions were causing issues with AWS Lambda, but we fixed it with a small patch). As far as I know, we are now trying to move towards a Dagster Op-Graph-based solution (instead of AWS StepFunctions). We still have a lot of open questions and a lot of things to learn about Dagster. I'm also confused about how pipes and op-graph-based solutions will work together.

I agree with you! An event-based/callback solution could be much better than a lot of processes with a single responsibility: polling an external resource to know if it has finished or not. It's a huge waste and not fault-tolerant (but perhaps I don't fully understand the concept of pipes yet).

**Comment by matthewwillian** [comment]
Hey folks super interested to see the outcome of this discussion. Would love to collaborate on a potential solution here if there's interest.
